### PR TITLE
[plugin] Remove LocalityFixme über hack.

### PR DESCRIPTION
**Pull request by ejgallego**
To that extent we introduce a new prototype vernacular extension
macro `VERNAC COMMAND FUNCTIONAL EXTEND` that will take a
function with the proper parameters and attributes, where we
start to encapsulate modifiers to vernac commands.

This is a continuation of #6183 and another step towards a more
functional interpretation of commands, and needs of course more
refinement, in particular we should move `vernac_command` to its
own file and make `Vernacentries` consistent wrt it.

### CODE REVIEW

File: API/API.mli [hunk 1]
```diff
@@ -5949,8 +5946,15 @@ sig
 
   type deprecation = bool
 
+  type atts = {
+    loc : Loc.t option;
+    locality : bool option;
```
Comment (ppedrot) [inline API/API.mli]: Not sure it is meaningful to special-case locality in a future world where we have attributes, as it's a part of an effect (e.g. local definition but not local instance and conversely). But I agree that we'll have to do that progressively.

File: grammar/vernacextend.mlp [hunk 2]
```diff
@@ -162,22 +166,27 @@ EXTEND
...
-      st -> st].
-
-      At some point We should solve the mess and extend
-      vernacextend.mlp with locality info. *)
   rule:
     [ [ "["; s = STRING; l = LIST0 args; "]";
         d = OPT deprecation; c = OPT classifier; "->"; "["; e = Pcaml.expr; "]" ->
       let () = if s = "" then failwith "Command name is empty." in
-      let b = <:expr< fun loc -> ( let () = $e$ in fun st -> st ) >> in
+      let b = <:expr< fun ~atts ~st -> ( let () = $e$ in st ) >> in
```
Comment (ppedrot) [inline grammar/vernacextend.mlp]: What is that piece of code below supposed to do? It doesn't seem to be related to localities.

Comment (ejgallego) [reply grammar/vernacextend.mlp]: You mean `fun_rule`, this is basically duplicate code, `rule` is an adaptor for imperative extensions, `fun_rule` is an adaptor for functional extension. That is to say, they are identical but for:
```
fun ~atts ~st -> ( let () = $e$ in st )
vs
$e$
```
[so `e` in the second case has type `unit` vs in the first which is a functional]

Help on how to better organize this is very welcome, my camlp5-fu is nul.

File: API/API.mli (hunk 1 above)
Comment (ejgallego) [reply API/API.mli]: Yeah this commit is not very ambitious in that regard as it doesn't mean to interfere with the people working on attributes but do the minimum to remove the `LocalityFixme` hack.

**Comment by ejgallego** [comment]
So this should be ready to go, modulo @ppedrot 's blessing.

Indeed we should figure out maybe a better way to specify extensions with different types [right now it is done via boilerplate], but here we need to follow @ppedrot 's lead obviously, in particular given the current plans.

**Comment by ejgallego** [comment]
Note that this depends on a LTAC2 overlay [kinda trivial]

**Review by ppedrot — APPROVED** [review]
LGTM.

File: API/API.mli (hunk 1 above)
Comment (maximedenes) [reply API/API.mli]: This first step is consistent with the attributes plan AFAICT.

**Review by maximedenes — APPROVED** [review]
Had only a superficial look, but it seems good.

File: vernac/locality.ml [hunk 3]
```diff
@@ -6,22 +6,12 @@
...
 
-open Pp
-
 (** * Managing locality *)
 
 let local_of_bool = function
   | true -> Decl_kinds.Local
   | false -> Decl_kinds.Global
 
-let check_locality locality_flag =
```
Comment (maximedenes) [inline vernac/locality.ml]: Wait! Where is this function gone?

Comment (maximedenes) [reply vernac/locality.ml]: @ejgallego IIUC, there is a bit of a design flaw in this PR: it seems that vernac command not supporting locality cannot use `VERNAC COMMAND FUNCTIONAL EXTEND`, or they will have to implement themselves the logic that prints an error message if a `Local` prefix is provided. Is that correct?

Comment (maximedenes) [reply vernac/locality.ml]: In fact, it is worse than that. Try for example:
```
Require Import FunInd.
Local Function f (n : nat) := n.
```

Previously, it was failing (admittedly, in a bit of a weird way) because `Function` does not support the `Local` prefix. Now it silently ignores `Local`, which is a regression.

Comment (maximedenes) [reply vernac/locality.ml]: And FTR, nothing is specific to `Function` here, it will be the same for all plugin-defined vernacular commands that don't support locality.

Comment (ppedrot) [reply vernac/locality.ml]: That doesn't seem wrong to me: in an attribute-based solution, it would be up to the client to check that all passed options make sense...

Comment (maximedenes) [reply vernac/locality.ml]: I hope it would be declarative, though.

Comment (maximedenes) [reply vernac/locality.ml]: I still think this is an ok intermediate state towards a better support of attributes, but I think it would be a bad idea to release 8.8 if such checks are not done at all, since they used to be done.

Comment (ejgallego) [reply vernac/locality.ml]: I was of the opinion of @ppedrot here, this PR didn't intend to solve the "attribute" problem, but the state problem, so it is up to the command to implement a warning.

I didn't want to put my hands into that as this is an area I thought would overlap with the attributes work. On the other hand I think there is a `command_supports_locality` function that should emit the warning, right? Plugins cannot register with it thou, but again I thought that people working on the attributes would handle that in their own PR.

**Comment by ejgallego** [comment]
Thanks!

File: grammar/vernacextend.mlp (hunk 2 above)
Comment (SkySkimmer) [inline grammar/vernacextend.mlp]: Camlp5 says:

>File "grammar/vernacextend.mlp", line 173, characters 27-32:
>Warning: syntax deprecated since version 6.00

File: vernac/locality.ml (hunk 3 above)
Comment (ejgallego) [reply vernac/locality.ml]: Regarding @maximedenes example, I think that if a command does not support locality, the parsing rule should not be indeed declared in the first place, so I am not sure that we can attribute the bad behavior in your example to the vernac interpretation infrastructure.

On a more general note, I think that indeed a possibility for attributes is for them to be reflect in the type of the generated rule. IMHO, dealing with this at the macro level should be better than using `Obj` hacks style. For example:

```ocaml
[local, pure] VERNAC EXTEND COMMAND foo ..
 { $e$ }
```
should require `e` to have type `?locality -> ~st -> args -> unit` etc... [using row types could also work.

Comment (Zimmi48) [reply vernac/locality.ml]: @maximedenes If you think that some action needs to be taken on this subject before the 8.8 release (and even if the attributes are on the roadmap), please open a new issue with the 8.8 milestone and a blocker label. Otherwise there is always a risk to forget.

Comment (herbelin) [reply vernac/locality.ml]: Looking at this PR: has something be done so that we don't release a version which stops silently dropping a locality flag, as in @maximedenes' example:
```coq
Require Import FunInd.
Local Function f (n : nat) := n.
```
Maybe should I open an issue?

Comment (maximedenes) [reply vernac/locality.ml]: Already done, see https://github.com/coq/coq/issues/6239

Comment (herbelin) [reply vernac/locality.ml]: Oups :(
